src/wrapper/jnienv.rs

```diff
@@ -1919,9 +1919,13 @@ impl<'local> JNIEnv<'local> {
         &mut self,
         obj: &'obj_ref JString<'other_local>,
     ) -> Result<JavaStr<'local, 'other_local, 'obj_ref>> {
-        let string_class = self.find_class("java/lang/String")?;
-        let obj_class = self.get_object_class(obj)?;
-        if !self.is_assignable_from(string_class, obj_class)? {
+        static STRING_CLASS: OnceCell<GlobalRef> = OnceCell::new();
```

Comment:
Hmm, interesting point, thanks.

Do you think we could make the lazy value a field of JNIEnv instead of static? That way it cannot be reused across VMs? I had some trouble with the `#[repr(transparent)]` requirement here, not sure the best way to implement the idea.
Or could we reset the value whenever a new JNIEnv is initialized? Not sure about the safety here, just an initial thought.

Reply:
I just saw in https://docs.oracle.com/en/java/javase/17/docs/specs/jni/invocation.html#jni_createjavavm:

> Creation of multiple VMs in a single process is not supported.

So I don't think we need to worry about Create->Destroy->Create.

Reply:
Storing fields in `JNIEnv` is not possible, because it is passed directly from the JVM upon calling a native method.

I had interpreted the JNI specification's statement as meaning it's not allowed to create multiple *concurrent* VMs. But, according to the output of this program, it looks like create→destroy→create isn't allowed, either.

```toml
# Cargo.toml

[package]
name = "multijvm"
edition = "2021"

[dependencies]
jni = { version = "0.21.1", features = ["invocation"] }
```

```rust
// main.rs

use jni::JavaVM;

fn new_jvm() -> JavaVM {
	JavaVM::new(
		jni::InitArgsBuilder::new()
		.version(jni::JNIVersion::V8)
		.build()
		.unwrap()
	).unwrap()
}

fn main() {
	let jvm = new_jvm();

	unsafe {
		jvm.destroy().unwrap();
	}

	let jvm = new_jvm();

	let mut env = jvm.attach_current_thread_permanently().unwrap();
	let system_out = env.get_static_field("java/lang/System", "out", "Ljava/io/PrintStream;").unwrap().l().unwrap();
	let message = env.new_string("Hello, world!").unwrap();
	env.call_method(&system_out, "println", "(Ljava/lang/String;)V", &[(&message).into()]).unwrap();
}
```

Output:

```
thread 'main' panicked at src/main.rs:9:7:
called `Result::unwrap()` on an `Err` value: Create(JniCall(Unknown))
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

If the `jvm.destroy()` and second `new_jvm()` calls are commented out, the program prints `Hello, world!` successfully.

So, if the JVM is storing global state, I guess we can do it too. 🤷‍♂️

Reply:
Thanks :). Yes, at least in hotspot it's explicitly [denied](https://github.com/openjdk/jdk17u/blob/5fb5ba651db3f6428916ad9e49cf5ff3faaed509/src/hotspot/share/prims/jni.cpp#L3605-L3607
): 

> If a previous creation attempt succeeded and we then
  // destroyed that VM, we will be prevented from trying to recreate
  // the VM in the same process, as the value will still be 0.

So do you think my approach can be approved? And are you a maintainer that can approve the workflow?

Reply:
No, sorry. I'm only commenting. @rib is the maintainer.